datafusion/core/src/physical_optimizer/enforce_sorting.rs

```diff
@@ -1049,6 +1065,98 @@ mod tests {
+            ],
+            input,
+        );
+
+        let expected_input = [
+            "SortExec: expr=[non_nullable_col@1 ASC,nullable_col@0 ASC], preserve_partitioning=[false]",
+            "  SortExec: TopK(fetch=2), expr=[non_nullable_col@1 ASC], preserve_partitioning=[false]",
+            "    MemoryExec: partitions=1, partition_sizes=[0]",
+        ];
+        let expected_optimized = [
+            "SortExec: TopK(fetch=2), expr=[non_nullable_col@1 ASC,nullable_col@0 ASC], preserve_partitioning=[false]",
+            "  MemoryExec: partitions=1, partition_sizes=[0]",
+        ];
+        assert_optimized!(expected_input, expected_optimized, physical_plan, true);
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_remove_unnecessary_sort7() -> Result<()> {
+        let schema = create_test_schema()?;
+        let source = memory_exec(&schema);
+        let input = Arc::new(SortExec::new(
+            vec![sort_expr("non_nullable_col", &schema)],
+            source,
+        ));
+        let limit = Arc::new(LocalLimitExec::new(input, 2));
+        let physical_plan = sort_exec(
+            vec![
+                sort_expr("non_nullable_col", &schema),
+                sort_expr("nullable_col", &schema),
+            ],
+            limit,
+        );
+
+        let expected_input = [
+            "SortExec: expr=[non_nullable_col@1 ASC,nullable_col@0 ASC], preserve_partitioning=[false]",
+            "  LocalLimitExec: fetch=2",
+            "    SortExec: expr=[non_nullable_col@1 ASC], preserve_partitioning=[false]",
+            "      MemoryExec: partitions=1, partition_sizes=[0]",
+        ];
+        let expected_optimized = [
+            "LocalLimitExec: fetch=2",
+            "  SortExec: TopK(fetch=2), expr=[non_nullable_col@1 ASC,nullable_col@0 ASC], preserve_partitioning=[false]",
+            "    MemoryExec: partitions=1, partition_sizes=[0]",
+        ];
+        assert_optimized!(expected_input, expected_optimized, physical_plan, true);
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_do_not_pushdown_through_limit() -> Result<()> {
+        let schema = create_test_schema()?;
+        let source = memory_exec(&schema);
+        // let input = sort_exec(vec![sort_expr("non_nullable_col", &schema)], source);
+        let input = Arc::new(SortExec::new(
+            vec![sort_expr("non_nullable_col", &schema)],
+            source,
+        ));
+        let limit = Arc::new(GlobalLimitExec::new(input, 0, Some(5))) as _;
+        let physical_plan = sort_exec(vec![sort_expr("nullable_col", &schema)], limit);
+
+        let expected_input = [
+            "SortExec: expr=[nullable_col@0 ASC], preserve_partitioning=[false]",
+            "  GlobalLimitExec: skip=0, fetch=5",
+            "    SortExec: expr=[non_nullable_col@1 ASC], preserve_partitioning=[false]",
+            "      MemoryExec: partitions=1, partition_sizes=[0]",
+        ];
+        let expected_optimized = [
```

Comment:
I don't understand why this plan doesn't push the limit into the sort as well like it does in the plan above

Like why is it not

```
            "LocalLimitExec: fetch=2",
            "  SortExec: TopK(fetch=2), expr=[non_nullable_col@1 ASC], preserve_partitioning=[false]",
            "    MemoryExec: partitions=1, partition_sizes=[0]",
```

(is it the diifference between GlobalLimitExec and LocalLimitExec 🤔 )

Reply:
The reason is that at the input plan: 
```
            "SortExec: expr=[nullable_col@0 ASC], preserve_partitioning=[false]",
            "  GlobalLimitExec: skip=0, fetch=5",
            "    SortExec: expr=[non_nullable_col@1 ASC], preserve_partitioning=[false]",
            "      MemoryExec: partitions=1, partition_sizes=[0]",
```
The expressions in the bottom `SortExec` (which is `non_nullable_col@1 ASC`) and top `SortExec` (which is `nullable_col@0 ASC`) are different and not compatible. This situation might arise with following kind of query
```
SELECT * FROM (SELECT * FROM table ORDER BY a ASC LIMIT 10) ORDER BY b ASC;
```
User might want to take subset of a table with a given order. Then might want to work on that table . Hence pushing down sort in this case would produce different result than the original query.

Reply:
oh I didn't see the sorts were on different keys -- thank you

I do wonder if the limit could be pushed into the lower sort, like

```rust
            "SortExec: expr=[nullable_col@0 ASC], preserve_partitioning=[false]",
            "  GlobalLimitExec: skip=0, fetch=5",
            "    SortExec: TopK(fetch=5) expr=[non_nullable_col@1 ASC], preserve_partitioning=[false]", <--- add TopK(fetch) here
            "      MemoryExec: partitions=1, partition_sizes=[0]",
```

Reply:
Yes, the limit pushdown rule will take care of it.